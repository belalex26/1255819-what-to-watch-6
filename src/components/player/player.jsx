import React, {useEffect, useRef, useState} from 'react';
import {Link} from 'react-router-dom';
import {connect} from 'react-redux';
import {useParams} from "react-router-dom";
import PropTypes from 'prop-types';

import shapeOfFilm from '../../proptypes/shape-of-film';
import LoadingScreen from '../loading-screen/loading-screen';
import PlayButton from '../play-button/play-button';
import PlayerToggler from '../player-toggler/player-toggler';

import {fetchFilmById} from '../../store/api-actions';
import {getFilm, getFilmLoadedStatus} from '../../store/selectors';
import {format} from '../../utils';


const Player = (props) => {
  const {film, isFilmLoaded, onLoad} = props;

  const id = parseInt(useParams().id, 10);

  const [isPlaying, setIsPlaying] = useState(false);
  const [isFullScreen, setIsFullScreen] = useState(false);
  const [timeLapse, setTimeLapse] = useState(false);
  const [previousTime, setPreviousTime] = useState(0);
  const [togglerProgress, setTogglerProgress] = useState(0);

  const videoRef = useRef();

  const hrefToFilmPage = `/films/${film.id}`;

  useEffect(() => {
    onLoad(id);
  }, [id]);

  useEffect(() => {
    if (!isFilmLoaded) {
      return;
    }
    if (isPlaying) {
      videoRef.current.play();
      return;
    }

    videoRef.current.pause();
  }, [isPlaying, isFilmLoaded]);

  if (!isFilmLoaded) {
    return (
      <LoadingScreen />
    );
  }

  const handlePlayBtnClick = () => {
    setIsPlaying(!isPlaying);
  };

  const handleFullScreenBtnClick = () => {
    if (!isFullScreen) {
      videoRef.current.requestFullscreen();
      setIsFullScreen(true);
    } else {
      videoRef.current.exitFullscreen();
      setIsFullScreen(false);
    }
  };

  const handleEndVideo = () => {
    videoRef.current.pause();
    videoRef.current.currentTime = 0;
    setIsPlaying(!isPlaying);
  };

  const handleTimeUpdate = () => {
    const duration = videoRef.current.duration;
    const currentTime = videoRef.current.currentTime;
    const time = Math.floor((duration - currentTime) % 60);

    if (previousTime !== time) {
      setPreviousTime(time);
      setTimeLapse(format(duration - currentTime));
    }
    const progress = Math.round((currentTime / duration * 100) * 100) / 100;
    setTogglerProgress(progress);
  };

  return (
    <div className="player">
      <video
        muted={false}
        src={film.videoLink}
        ref={videoRef}
        className="player__video"
        poster={film.backgroundImage}
        onTimeUpdate={handleTimeUpdate}
        onEnded={handleEndVideo}
      >
      </video>
      <Link to={hrefToFilmPage}>
        <button type="button" className="player__exit">Exit</button>
      </Link>

      <div className="player__controls">
        <div className="player__controls-row">
          <PlayerToggler togglerProgress={togglerProgress}/>
          <div className="player__time-value">{timeLapse}</div>
        </div>

        <div className="player__controls-row">
          <PlayButton
            isPlaying={isPlaying}
            onButtonClick={handlePlayBtnClick}
          />
          <div className="player__name">{film.name}</div>
          <button
            type="button"
            className="player__full-screen"
            onClick={handleFullScreenBtnClick}
            data-testid="full-screen_btn"
          >
            <svg viewBox="0 0 27 27" width={27} height={27}>
              <use xlinkHref="#full-screen" />
            </svg>
            <span>Full screen</span>
          </button>
        </div>
      </div>
    </div>
  );
};

Player.propTypes = {
  film: shapeOfFilm(),
  isFilmLoaded: PropTypes.bool.isRequired,
  onLoad: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  film: getFilm(state),
  isFilmLoaded: getFilmLoadedStatus(state),
});

const mapDispatchToProps = (dispatch) => ({
  onLoad(id) {
    dispatch(fetchFilmById(id));
  },
});

export {Player};
export default connect(mapStateToProps, mapDispatchToProps)(Player);
