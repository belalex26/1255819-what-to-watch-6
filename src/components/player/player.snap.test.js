import React from 'react';
import {Router} from 'react-router-dom';
import {Provider} from 'react-redux';
import {createMemoryHistory} from 'history';
import {render} from '@testing-library/react';
import configureStore from 'redux-mock-store';

import {Player} from './player';
import {mockFilm} from '../../test-mocks';

const mockStore = configureStore({});
const store = {
  FILMS: {
    film: mockFilm,
  }
};

it(`Player renders correctly`, () => {
  const history = createMemoryHistory();
  const {container} = render(
      <Provider store={mockStore(store)}>
        <Router history={history}>
          <Player
            film={mockFilm}
            isFilmLoaded={true}
            onLoad={jest.fn()}
          />
        </Router>
      </Provider>
  );

  expect(container).toMatchSnapshot();
});

