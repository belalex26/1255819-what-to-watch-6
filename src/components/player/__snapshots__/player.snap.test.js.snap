// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player renders correctly 1`] = `
<div>
  <div
    class="player"
  >
    <video
      class="player__video"
      poster="https://i.picsum.photos/id/604/248/152.jpg?hmac=Yu4rnPqlL8pPa8b7cdPSosfFApQXtQmy56BW52elqto"
      src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
    />
    <a
      href="/films/6"
    >
      <button
        class="player__exit"
        type="button"
      >
        Exit
      </button>
    </a>
    <div
      class="player__controls"
    >
      <div
        class="player__controls-row"
      >
        <div
          class="player__time"
        >
          <progress
            class="player__progress"
            max="100"
            value="0"
          />
          <div
            class="player__toggler"
            style="left: 0%;"
          >
            Toggler
          </div>
        </div>
        <div
          class="player__time-value"
        />
      </div>
      <div
        class="player__controls-row"
      >
        <button
          class="player__play"
          type="button"
        >
          <svg
            height="19"
            viewBox="0 0 19 19"
            width="19"
          >
            <use
              xlink:href="#play-s"
            />
          </svg>
          <span>
            Play
          </span>
        </button>
        <div
          class="player__name"
        >
          The Exorcist
        </div>
        <button
          class="player__full-screen"
          data-testid="full-screen_btn"
          type="button"
        >
          <svg
            height="27"
            viewBox="0 0 27 27"
            width="27"
          >
            <use
              xlink:href="#full-screen"
            />
          </svg>
          <span>
            Full screen
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
